### PR TITLE
issue fix #1801 SKU filter issue fixed.

### LINKED ISSUE #1801: REST API: Issue when sorting the variation products via SKU in products analytics tabs in single product
**Describe the bug**
When We have multiple orders of the same product with different variation and that both products have different SKUs at that time there is a bug in REST API so when We click on sort by SKU on the product single page on analytics at that time it gives 400 error because there is not SKU parameter in REST API.
filter=single_product page.

File wc-includes/api/class-wc-admin-rest-reports-variations-controller.php

Line 287:
$params['orderby']       = array(
			'description'       => __( 'Sort collection by object attribute.', 'wc-admin' ),
			'type'              => 'string',
			'default'           => 'date',
			'enum'              => array(
				'date',
				'net_revenue',
				'orders_count',
				'items_sold',
			),
			'validate_callback' => 'rest_validate_request_arg',
		);

It does not take sku option from order by params so it gives 400 error in the rest API response.
wc-admin

Also, the query is not set for the filter by SKU in variation in the data store files.

**To Reproduce**

1. Create a product with 2 or more variation or choose the product which has 2 or more variations.
2. Assign different SKUs to all variants
2. Place different orders of that same product with different variations.
3. Go, WooCommerce admin on analytics menu and click on products
4. Where you will see the product name(which you have ordered with different variation) and Item sold columns 
screenshot-localhost-2019 03 14-12-11-09

5. Click on that product
6. You will be able to see all products with variation
Like
screenshot-localhost-2019 03 14-14-37-46

7. Now you will be able to see both variations have different SKUs.
8. Click on SKU column to sort it.
9. You will see the following error,
ezgif com-video-to-gif (5)

10. The columns will become empty and in REST API response it wil return 400 error.
**Expected behavior**
It should Filter according to SKU in Ascending, descending flow.

**Screenshots**
ezgif com-video-to-gif (5)

**Desktop (please complete the following information):**

* OS: Ubuntu
* Browser Chrome

**Additional context**
I have fixed it and attached a PR on it.

### PR DESCRIPTION
**Pull request by ronakganatra9**
Fixes #1801 #1802 #1808 

Attached a new PR

### CODE REVIEW

**Review by psealock — COMMENTED** [review]
Thanks for cleaning up the commit history with this one @ronakganatra9.

When I tested this, I needed to add `isSortable: true,` to this block to be able to sort the SKU column.

https://github.com/woocommerce/woocommerce-admin/blob/0f20052de6f9b706baaccfc6f8cb64c713c53d37/client/analytics/report/products/table-variations.js#L39-L43

Tests are also failing. One issue has already been noted https://github.com/woocommerce/woocommerce-admin/issues/1828. The other is a tabs v. spaces issue. 

I don't know why the pre-commit checks didn't pick up on this, but you can run the php codesniffer to automatically fix these:

```
 ./vendor/bin/phpcbf includes/data-stores/class-wc-admin-reports-variations-data-store.php includes/api/class-wc-admin-rest-reports-variations-controller.php
```

**Comment by ronakganatra9** [comment]
Hello @psealock ,

I have implemented the changes and fixed the PHPCS errors and warnings.

File: includes/data-stores/class-wc-admin-reports-variations-data-store.php [hunk 1]
```diff
@@ -127,6 +158,9 @@ protected function normalize_order_by( $order_by ) {
 		if ( 'date' === $order_by ) {
 			return $table_name . '.date_created';
 		}
+		if ( 'sku' === $order_by ) {
+			return 'postmeta.meta_value';
```
Comment (psealock) [inline includes/data-stores/class-wc-admin-reports-variations-data-store.php]: `WC_Admin_Reports_Products_Data_Store` uses just `meta_value` here. I don't think it matters, but any reason not to use the same?

**Review by psealock — COMMENTED** [review]
Looking good @ronakganatra9 !

Just a small comment. A rebase to combine your commits into one and this is good to go.

Thanks again

File: includes/data-stores/class-wc-admin-reports-variations-data-store.php (hunk 1 above)
Comment (ronakganatra9) [reply includes/data-stores/class-wc-admin-reports-variations-data-store.php]: Sorry, @psealock I am not understood what you want to explain to me. Can you please elaborate it little more.

Comment (ronakganatra9) [reply includes/data-stores/class-wc-admin-reports-variations-data-store.php]: Hi @psealock , I have tried the same but it was not working if I was using only meta_value. SO, I hav updated as postmeta.metavalue.
As, Comparing to the code both the files have minor structure changes. And it is working fine with postmeta.meta_value also.

Let me know your thoughts on the same.

Thanks & Regards,
Ronak Ganatra

File: includes/data-stores/class-wc-admin-reports-variations-data-store.php [hunk 2]
```diff
@@ -78,11 +78,41 @@ public function __construct() {
...
+		global $wpdb;
+		$order_product_lookup_table = $wpdb->prefix . self::TABLE_NAME;
+
+		$sql_query['order_by_clause'] = '';
+		if ( isset( $query_args['orderby'] ) ) {
+			$sql_query['order_by_clause'] = $this->normalize_order_by( $query_args['orderby'] );
+		}
+
+		if ( 'postmeta.meta_value' === $sql_query['order_by_clause'] ) {
+			$sql_query['from_clause'] .= " JOIN {$wpdb->prefix}postmeta AS postmeta ON {$order_product_lookup_table}.variation_id = postmeta.post_id AND postmeta.meta_key = '_sku'";
```
Comment (Aljullu) [inline includes/data-stores/class-wc-admin-reports-variations-data-store.php]: I have been testing this when a comparison is active and couldn't get it working. You can reproduce it selecting a variable product, then selecting a _Comparison_ filter and selecting some of them:
image

This is a bit complex because the SQL command changes depending if a comparison is active or not. You can use `sku` sorting in the products table as a reference in case it helps:
https://github.com/woocommerce/woocommerce-admin/blob/master/includes/data-stores/class-wc-admin-reports-products-data-store.php#L135
Basically, you will have to make that join to be in the inner query when there is no comparison and to be in the outer query when there is a comparison.

I think the product endpoint is quite similar to this one, so taking a look at the file I linked, should be helpful. :+1:

**Review by Aljullu — COMMENTED** [review]
Thanks for working on this @ronakganatra9! I have to say this is a complex issue, but it's looking good. I posted some comments that might help you with the work here.

File: includes/data-stores/class-wc-admin-reports-variations-data-store.php (hunk 1 above)
Comment (ronakganatra9) [reply includes/data-stores/class-wc-admin-reports-variations-data-store.php]: Hi @psealock , Updated the changes.

File: includes/data-stores/class-wc-admin-reports-variations-data-store.php (hunk 2 above)
Comment (ronakganatra9) [reply includes/data-stores/class-wc-admin-reports-variations-data-store.php]: @Aljullu , Thanks for reviewing these changes and sharing your thoughts. I am looking for this solution till that what I suggest is that let's create a separate issue for this and merge these changes so that it can be less chance of conflicts.

Comment (Aljullu) [reply includes/data-stores/class-wc-admin-reports-variations-data-store.php]: @ronakganatra9 thank you for addressing the other issues!

I would say it's better to wait until the feature is complete before merging it into master. Otherwise, users might try sorting by `sku` with an active comparison and that would make Woocommerce Admin to crash.

As far as I know, there is nobody currently working in the same files you modified in this PR, so I don't expect you to end up with many conflicts. In any case, if you have issues with that, we can help in solving them.

Does that sound good?

Comment (ronakganatra9) [reply includes/data-stores/class-wc-admin-reports-variations-data-store.php]: @Aljullu , Comparision is not available on products analytics page so it is not occurring the issue there. But in variation page of single product, there is a comparison option available so I recommend to push these changes or create a different issue for that because that is a complex issue.

Comment (Aljullu) [reply includes/data-stores/class-wc-admin-reports-variations-data-store.php]: Thanks for your patience here @ronakganatra9, there were so many things going on last week I didn't have time to take another look at this until now.

I created a follow-up PR that will fix the issue with the comparison, so you don't have to worry about that, and we can proceed merging this one. :slightly_smiling_face: 

Thank you again for your help!

**Review by Aljullu — APPROVED** [review]

File: includes/data-stores/class-wc-admin-reports-variations-data-store.php (hunk 2 above)
Comment (ronakganatra9) [reply includes/data-stores/class-wc-admin-reports-variations-data-store.php]: Thank you so much @Aljullu for this.
